### DISCUSSION TITLE
Using PnP Core SDK to recreate PowerShell scripts on a .Net console app

### DISCUSSION THREAD

**Discussion by Barbarur** [pnp/pnpcore#1036] [Q&A] [answered]
I want to recreate on a .Net console app several PowerShell scripts I have.

While the PnP.PowerShell module is quite complete and straight forward. PnP Code SDK seems I have to recreate many things that are already done in the PowerShell module.

I'm not sure if I'm looking at the right module here for what I need.

The intention of doing this rather than sharing PowerShell scripts is to avoid asking user to use PS and install modules they don't know, which they are not familiar.
This .Net app would go straight of what they need.

Is this the correct package I should use for my purpose?

Thank you in advance.

**Comment by jansenbe** [comment]
@Barbarur : PnP PowerShell under the cover uses PnP Core SDK (this package), PnP Framework combined with a lot of other code that calls into SharePoint and Teams using SharePoint CSOM/REST and Microsoft Graph API calls. So it's possible to replace a PnP PowerShell script with a console application, how easy that is depends a lot on the which cmdlets the script that you're replacing was using.

**Comment by bagoo-cm** [comment]
@jansenbe, as complete Sharepont freshman, I was a bit puzzled of all the libraries and the best way to set things up.
Following on @Barbarur's question, we too succeeded in using powershell to connect to Sharepoint. We're using a freshly made Azure AD cetificate and connected through powershell with the four arguments: Connect-PnPOnline -**URL** .. -**tenant** .. -**clientID** .. -**Thumbprint** ..

Our goal is to connect from a .NET Framework 4.8 Windows service app. So far, I guess I just have to throw in a .NET Standard 2.0 component to make the ShP connection with and pass the ClientContext to the f4.8.

I ran by the Microsoft.SharePointOnline.CSOM - and the PnP.Framework libraries in search of an authentication method that takes these four arguments. No luck there.

After reading your comment here, I understand that the powerscript connection we're succeeding in, is based on the PnP.Core library.
Porting the ps cmdlets straight to our app isn't obvious in f4.8. If I dive into the PnP.Core library, will I find a/some class(es) that take the four arguments and return me a authenticated ClientContext?  

Your input will be most welcome as I done some diving already, the last few days ;-)
Thank you kindly!

**Comment by jansenbe** [comment, marked as answer]
@bagoo-cm : please check the code in the application permissions polyglot sample (https://github.com/pnp/pnpcore/tree/dev/docs/polyglot). It uses a certificate that it reads from a certificate store on the machine.

**Reply by bagoo-cm** [reply]
Thank you @jansenbe, for the tip of looking into the PnP.Core example.

For the time being, we remain in .NET Framework 4.8 with our main service application.
This main module expects a Microsoft.SharePoint.Client.ClientContext to be returned after authentication.

Unfortunately, the PnP.Core returns a PnPContext which I couldn't cast to ClientContext.
So, until we refactor the complete service, I found it convenient to use the AuthenticationManager from PnP.Framework instead.
Where:
`ClientContext _context = new PnP.Framework.AuthenticationManager( clientID, certificate, tenantID )
                                      .GetContext(siteUrl) ;`
